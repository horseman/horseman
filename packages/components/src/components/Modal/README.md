The modal is a full screen take over of the page. A modal can have several
different types of content within it.

There is only intended to be a single modal on the page at a time, and triggered
via redux.

The open / closed state is manually set by the `handleClose` prop

## Installation

Add the modal reducers to your reducers object

```js
import { reducers } from "@horseman/components";

export default combineReducers({
  ...reducers.modal,
});
```

Connect the modal to your redux store

```js
/* ConnectedModal.js */
import { connect } from "react-redux";

import { Modal, actions } from "@horseman/components";

export const mapStateToProps = state => ({
  isOpen: state.modal.modalType !== null,
  type: state.modal.modalType,
  contentProps: state.modal.props,
});

export const mapDispatchToProps = dispatch => ({
  handleClose: () => {
    dispatch(actions.modal.hideModal());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Modal);
```

