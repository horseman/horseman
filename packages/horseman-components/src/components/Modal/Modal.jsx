import React from "react";
import PropTypes from "prop-types";
import ReactModal from "react-modal";

const Modal = ({
  isOpen,
  mappings,
  type,
  contentProps,
  handleClose,
  closeButton,
}) => {
  const ContentComponent = mappings[type];
  const CloseComponent = closeButton;
  const content = type ? <ContentComponent {...contentProps} /> : null;
  return (
    <ReactModal
      shouldCloseOnOverlayClick
      contentLabel="Modal"
      isOpen={isOpen}
      onRequestClose={handleClose}
      style={{
        overlay: {
          alignItems: "center",
          backgroundColor: "rgba(0, 0, 0, .9)",
          display: "flex",
          justifyContent: "center",
          zIndex: 10,
        },
        content: {
          border: "none",
          borderRadius: "0",
          maxWidth: "900px",
          position: "static",
          width: "90%",
          zIndex: "3",
        },
      }}
    >
      {content}
      <CloseComponent onClick={handleClose} />
    </ReactModal>
  );
};

Modal.defaultProps = {
  isOpen: false,
  type: null,
};

Modal.propTypes = {
  /**
   * Determine whether or not to show the modal
   */
  isOpen: PropTypes.bool,

  /**
   * What happens when the 'x' button is clicked.
   */
  handleClose: PropTypes.func.isRequired,

  /**
   * Will determine the type of content that is rendered within the modal
   * itself.
   */
  type: PropTypes.string,

  /**
   * Props which will be passed to the underlying modal content component
   */
  contentProps: PropTypes.object,

  /**
   * Mapping of string type to the type of component that will be rendered
   */
  mappings: PropTypes.object.isRequired,

  /**
   * The component to be rendered in the close button spot
   */
  closeButton: PropTypes.node.isRequired,
};

export default Modal;
