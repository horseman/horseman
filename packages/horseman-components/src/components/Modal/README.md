The modal is a full screen take over of the page. A modal can have several
different types of content within it.

There is only intended to be a single modal on the page at a time, and triggered
via redux.

The open / closed state is manually set by the `handleClose` prop
