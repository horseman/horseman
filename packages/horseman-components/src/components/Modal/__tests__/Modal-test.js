import React from "react";
import { shallow } from "enzyme";

import Modal from "../";

describe("Modal", () => {
  const data = {
    handleClose: () => {},
    mappings: {
      div: <div>div</div>,
      span: <span>span</span>,
    },
    contentProps: {
      foo: "bar",
    },
    closeButton: <div>close</div>,
  };
  test("renders correctly", () => {
    const wrapper = shallow(<Modal {...data} />);
    expect(wrapper).toMatchSnapshot();
  });
});
