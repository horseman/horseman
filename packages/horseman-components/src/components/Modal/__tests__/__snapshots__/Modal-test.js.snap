// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={0}
  contentLabel="Modal"
  isOpen={false}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
  style={
    Object {
      "content": Object {
        "border": "none",
        "borderRadius": "0",
        "maxWidth": "900px",
        "position": "static",
        "width": "90%",
        "zIndex": "3",
      },
      "overlay": Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, .9)",
        "display": "flex",
        "justifyContent": "center",
        "zIndex": 10,
      },
    }
  }
>
  <closeButton
    onClick={[Function]}
  />
</Modal>
`;
